dsc/workflows/opencourseware.py

```diff
@@ -0,0 +1,183 @@
+            except FileNotFoundError:
+                bitstreams_without_metadata.append(item_identifier)
+
+        if any(bitstreams_without_metadata):
+            reconcile_error_message = {
+                "note": "Failed to reconcile bitstreams and metadata.",
+                "bitstreams_without_metadata": {
+                    "count": len(bitstreams_without_metadata),
+                    "identifiers": bitstreams_without_metadata,
+                },
+            }
+            logger.error(json.dumps(reconcile_error_message))
+            raise ReconcileError(json.dumps(reconcile_error_message))
+
+        logger.info(
+            "Successfully reconciled bitstreams and metadata for all "
+            f"items (n={len(item_identifiers)})."
+        )
+
+    def item_metadata_iter(self) -> Iterator[dict[str, Any]]:
+        """Yield source metadata from metadata JSON file in the zip file.
+
+        The item identifiers are retrieved from the filenames of the zip
+        files, which follow the naming format "<item_identifier>.zip".
+        """
+        s3_client = S3Client()
+        for file in s3_client.files_iter(
+            bucket=self.s3_bucket, prefix=self.batch_path, file_type=".zip"
+        ):
+            zip_file = f"s3://{self.s3_bucket}/{file}"
+            item_identifier = file.split("/")[-1].removesuffix(".zip")
+            yield {
+                "item_identifier": item_identifier,
+                **self._extract_metadata_from_zip_file(zip_file, item_identifier),
+            }
+
+    def _extract_metadata_from_zip_file(
+        self, file: str, item_identifier: str
+    ) -> dict[str, str]:
+        """Yield source metadata from metadata JSON file in zip archive.
+
+        This method expects a JSON file called "data.json" at the root
+        level of the the zip file.
+        """
+        with smart_open.open(file, "rb") as file_input, zipfile.ZipFile(
+            file_input
+        ) as zip_file:
+            for filename in zip_file.namelist():
+                metadata_json = f"{item_identifier}/data.json"
+                if filename == metadata_json:
+                    return self._read_metadata_json_file(zip_file, metadata_json)
+            raise FileNotFoundError(
+                "The required file 'data.json' file was not found in the zip file: "
+                f"{file}"
+            )
+
+    def _read_metadata_json_file(
+        self, zip_file: zipfile.ZipFile, metadata_json: str
+    ) -> dict[str, str]:
+        """Read source metadata JSON file."""
+        with zip_file.open(metadata_json) as file:
+            source_metadata = json.load(file)
+            source_metadata["instructors"] = self._get_instructors_delimited_string(
+                source_metadata["instructors"]
+            )
+            return source_metadata
+
+    def _get_instructors_delimited_string(self, instructors: list[dict[str, str]]) -> str:
```

Comment:
Noting @jonavellecuerdo's [comment #2 here](https://github.com/MITLibraries/dspace-submission-composer/pull/91#issuecomment-2615913965), around exploring if multi-value lists/arrays can be passed for metadata parsing, to avoid creating a delimited string just to parse it again.

I would be in favor of either an issue or a new ticket and _not_ incorporate that in this PR.  But noting here, as I think this PR exposes a situation when we're creating delimited strings from structured data only to re-structure it moments later.

Reply:
Agree that a ticket would be the best approach

Reply:
Ticket has been created [IN-1159](https://mitlibraries.atlassian.net/jira/software/c/projects/IN/boards/238?selectedIssue=IN-1159). :)